### DISCUSSION TITLE
Does pre-create access key offer better performance when we query a json object?

### DISCUSSION THREAD

**Discussion by Jack012a** [nlohmann/json#3861] [Q&A] [answered]
```
#include <iostream>
#include <iomanip>
#include <string_view>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

int main()
{
    // create a JSON object
    json object =
    {
        {"one", 1}, {"two", 2}, {"three", 2.9}
    };

    // query element with const char* key "two"
    // method A
    for(int i=0; i<1000; ++i)
    {
      std::cout << object["two"] << "\n\n";
    }
    
    // output element with string_view key "two"
    // method B
    static constexpr std::string_view key_two = "two"sv;
    for(int i=0; i<1000; ++i)
    {
      std::cout << object[key_two] << "\n\n";
    }
}
```

**Question**> Do we get better performance if we pre-allocate the key as I did in method B?

Thank you

**Comment by nlohmann** [comment]
If it does not make a difference for `std::map`, it should also make no difference here. Did you measure it?

**Comment by Jack012a** [comment, marked as answer]
Hello,

Based on my following test, each time a new temporary object is created. So the method TWO should be more efficient.
But I am not sure whether this demo exactly shows how std::map works.

```
#include <iostream>
#include <string>

struct ABC
{
    ABC(const char* name)
    {
        std::cout << "ABC(): " << name << std::endl;
    }
};

void function1(const ABC& abc)
{
    std::cout << "function1: " << std::endl;
}

int main()
{
    for(int i=0; i<3; ++i)
    {
        function1("abc");
    }

    return 0;
}
```

https://godbolt.org/z/55edovsGq

Output:
```
ABC(): abc
function1: 
ABC(): abc
function1: 
ABC(): abc
function1: 
```

**Reply by nlohmann** [reply]
Which version are you using? We recently overworked all operators that take a string literal to avoid temporaries.

**Reply by Jack012a** [reply]
Hello,

I have the version of nlohmann.v3.11.2 which is the latest version.

Does it mean I can just use the following syntax without creating a temporary key every time I call it?

` object["two"]`

Thank you

**Reply by nlohmann** [reply]
Yes.

**Reply by nlohmann** [reply]
Do you need further assistance with this issue?

**Reply by Jack012a** [reply]
no. Thank you!
